Parameterize GCSClient chunksize for better control when uploading or downloading data

geowurster opened the pull request:
Closes #1218 

I upped the default `chunksize` to `10 mb` and parameterized as `GCSClient(chunksize=10 * 1024 * 1024)`.

Creating `GCSClient()` instances can be expensive if you hit the API at the wrong time, so the user can also override the chunksize when calling the `download()` or `put()` methods directly rather than instantiating another instance of `GCSClient()` with a different chunk size.

@erikbern ready for review.  I didn't see a docstring convention mentioned anywhere so I wasn't sure how to format.

erikbern commented on luigi/contrib/gcs.py:
```diff
@@ -104,7 +107,9 @@ class GCSClient(luigi.target.FileSystem):
     """
-    def __init__(self, oauth_credentials=None, descriptor='', http_=None):
+    def __init__(self, oauth_credentials=None, descriptor='', http_=None,
+                 chunksize=10 * 1024 * 1024):
```
you should default this to CHUNKSIZE and set CHUNKSIZE to 10*1024*1024

Tarrasch commented:
Do @mikekap want to quickly review this too? This looks good to me. Just see Erik's comment. :)

geowurster commented:
@erikbern @Tarrasch Moved default chunksize to global variable.

erikbern commented:
LGTM

I realize maybe we should have made chunksize configurable but it's no biggie. let's merge

mikekap commented:
LGTM as well.

A small note - I didn't expect making a GCSClient to be a heavy op. I don't use the default client myself (to pass in a local descriptor, I have a globally memorized one). Perhaps if the default construction is that expensive, we should memorize it in the BigqueryTarget constructor?

geowurster commented:
@mikekap It's only heavy because of the [API discovery](https://github.com/spotify/luigi/blob/master/luigi/contrib/gcs.py#L113).  If you hit the Google API at a time with high latency then it can be expensive to build a `GCSClient()`.  Instantiating a client with every target creates more opportunities for encountering latency.

A generic luigi solution might be to just instantiate a `GCSClient()` in a `gcs.GCS_CLIENT` global variable and have the `GCSTarget()` grab that one rather than creating a new instance if a client is not supplied.

Our solution was to attach a singleton `GCSClient()` to a subclassed `luigi.task.Config` object paired with a `DynamicTarget()` that always gives us GCS targets with the instantiated client or a more appropriate target based on the `schema://`.  It also gave us the opportunity to create a `GCSClientStub()` that translates GCS paths to local paths for running locally and testing.  We have not started integrating `BigqueryTargets()` but will before too long and will report back if we find some latency.  The `BenthospConfig()` object always references a single client in a class variable no matter how many times it is instantiated.

```python
def DynamicTarget(path, **kwargs):

    """
    A single interface for dynamically retrieving luigi target objects based on
    the input path's schema.

        file://path = LocalTarget
        gs://path = GCSTarget

    The ``BenthospConfig()`` object is used to determine if a stubbed client
    should be used.

    Parameter
    ---------
    path : str
        Behaves just like ``path`` for any other target, except it requires a
        schema.
    kwargs : **kwargs, optional
        Additional keyword arguments for the desired target.
    """

    cfg = settings.BenthospConfig()

    if path.startswith('gs://'):
        kwargs.update(client=cfg.gcs_client)
        return gcs.GCSTarget(path, **kwargs)

    elif path.startswith('file://'):
        return luigi.LocalTarget(path[7:], **kwargs)

    else:
        raise ValueError("Unrecognized schema: {}".format(path))
```

```python
class GCSClientStub(object):

    """
    Drop-in replacement for luigi.contrib.gcs.GCSClient
    Maps a given GCS path to a given local file path so that any request using a gcs path
    will be executed on the local file system
    """

    def __init__(self, local_root=os.getcwd(), **kwargs):

        """
        Create a special GCS client that operates locally.
        Parameters
        ----------
        local_root : str, optional
            Consider this location to be the root of GCS.  Any sub-directories
            will be considered buckets.  Creating a bucket creates a sub-dir in
            this location.
        kwargs : **kwargs, (does nothing)
            Quietly eat arguments that would be passed to ``GCSClient()`.
        """

        local_root = os.path.abspath(local_root)
        if not os.path.exists(local_root):
            os.makedirs(local_root)

        self._local_root = local_root

        self.fs = LocalFileSystem()

    def _gcs2local(self, path):

        """
        Convert a GCS path to a local path
        """

        split = urlparse.urlsplit(path)
        bucket = posixpath.join(self._local_root, split.netloc)
        path = posixpath.join(bucket, split.path[1:])  # Slice off leading slash
        if not os.path.exists(bucket):
            os.makedirs(bucket)
        return path

    def _local2gcs(self, path):
        path = path.replace(self._local_root, '', 1)[1:]  # Slice off leading slash
        return 'gs://' + path

    def _path_to_bucket_and_key(self, path):
        (scheme, netloc, path, _, _) = urlsplit(path)
        assert scheme == 'gs'
        path_without_initial_slash = path[1:]
        return netloc, path_without_initial_slash


    def copy(self, source_path, destination_path):
        source_path = self._gcs2local(source_path)
        destination_path = self._gcs2local(destination_path)
        LocalTarget(source_path).copy(destination_path)

    def download(self, path):
        # GCSClient() downloads the file to a temp location and returns an open
        # file pointer, so we do the same thing here

        path = self._gcs2local(path)
        assert os.path.exists(path), "File does not exist: {}".format(path)
        with tempfile.NamedTemporaryFile(delete=False) as fp:
            LocalTarget(path).copy(fp.name)
            return gcs._DeleteOnCloseFile(fp.name, 'r')

    def exists(self, path):
        return self.fs.exists(self._gcs2local(path))

    def isdir(self, path):
        return self.fs.isdir((self._gcs2local(path)))

    def listdir(self, path):
        lpaths = self.fs.listdir(self._gcs2local(path))
        return (self._local2gcs(p)for p in lpaths)

    def mkdir(self, path, **kwargs):
        return self.fs.mkdir(self._gcs2local(path), **kwargs)

    def put(self, filename, dest_path, **kwargs):
        LocalTarget(filename).copy(self._gcs2local(dest_path))

    def put_string(self, contents, dest_path, **kwargs):
        dest_path = self._gcs2local(dest_path)
        bname = os.path.basename(dest_path)
        if not os.path.exists(bname):
            os.makedirs(bname)
        with LocalTarget(dest_path).open('w') as f:
            f.write(contents)

    def remove(self, path, **kwargs):
        # GCSClient does not fail on remove() if the path does not exist, but LocalFileSystem does,
        # so check to make sure it exists before attempting to remove
        if self.exists(path):
            return self.fs.remove(self._gcs2local(path), **kwargs)
        else:
            return False

    def rename(self, source_path, destination_path):
        source_path = self._gcs2local(source_path)
        destination_path = self._gcs2local(destination_path)
        LocalTarget(source_path).move(destination_path)
```

Tarrasch commented:
Thanks for the patch. I guess when users find that the value should be parametrized they'll add it :)
